### DISCUSSION TITLE
I have a combobox in a one-input form, and I need keydown.enter to submit the form.

### DISCUSSION THREAD

**Discussion by aseroff** [josefarias/hotwire_combobox#171] [Q&A]
It looks like deep in [navigate](https://github.com/josefarias/hotwire_combobox/blob/eae73f0c8b0bb453b3c0f3dfae4084b6f7c7f01f/app/assets/javascripts/hw_combobox/models/combobox/navigation.js#L33) Enter is bound to closing and blurring the combobox. Based on [this](https://github.com/josefarias/hotwire_combobox/blob/eae73f0c8b0bb453b3c0f3dfae4084b6f7c7f01f/app/presenters/hotwire_combobox/component.rb#L330) I figured I could add my own binding to enter and it would be prepended to the list of actions on the input, but it's not working.

`= form.combobox :search, library_textbooks_path, name_when_new: 'term', render_in: { partial: 'library/textbooks/search_result' }, dialog_label: 'Textbook Search', placeholder: 'Search the Library', data: { action: 'keydown.enter->form#submit' }`

Am I doing something wrong? If so, how would I accomplish my enter to submit use case?

**Comment by josefarias** [comment]
Hey @aseroff thanks for the question. I think your best bet will be to use a stimulus controller to auto-submit the form based on the events emitted by the combobox. Here's an example: https://github.com/josefarias/hotwire_combobox/issues/137#issuecomment-2035555941

Let me know if that doesn't work for you. Closing for now.

**Reply by aseroff** [reply]
Thanks as always for the fast response.

I'm not sure this solves my problem. When I hit Enter, a hw-combobox:selection event is fired, which is the same as when you click on a result. I need clicking on a result to do one thing, and pressing enter to do another. Is it possible to differentiate between the two?

**Reply by aseroff** [reply]
I've implemented a marginal solution using your approach but have two issues:

1. My handleSelection code sucks. Selecting from the combobox sets the value to a numerical id, which I'm using to differentiate between a click and an Enter keypress. (`if (isNaN(value))` 🤢). And the undesirable side effect of this is that searching for a numerical value will falsely trigger the execution of my click code instead of my enter code.
2. If you type fast enough and hit enter, you can beat the updating of the hidden field's value, resulting in a search of empty string when you previously had text in the input.

I feel that it may be necessary to make handling this use case a bit more first-class.

**Reply by aseroff** [reply]
For what it's worth, I managed to accomplish what I needed to using old fashioned event listeners and mutation observers, sans Stimulus. So I'm happy to continue discussing, but also my issue is resolved and no longer require any assistance, if you don't want to.

**Reply by buncis** [reply]
@josefarias @aseroff I tried to submit a form after user make a selection
here is my code 

```
<%= form.combobox :member_id, combobox_members_path, data: { action: "hw-combobox:selection->form#submit"} %>
```
but the form is not submitted, did I need to make a stimulus controller for form? or what?


edit : 

ya I need to execute the submit form from another controller

**Comment by LouisKottmann** [comment]
I have the same need to submit the form on pressing enter.

I found that this (now merged) #240 PR add the possibility to do this:

```
  <%= f.combobox 
             ...,
             input: {
               data: {
                 action: "keypress.enter->auto-submit#submit"
               }
             } %>
```

(And so I finally grasped the use of `customize`).

This produces, at the time of writing this comment, a combobox with an input that has this:

```
data-action="click->hw-combobox#toggle keydown->hw-combobox#prepareToFilter input->hw-combobox#filterAndSelect keydown->hw-combobox#navigate click@window->hw-combobox#closeOnClickOutside focusin@window->hw-combobox#closeOnFocusOutside turbo:before-stream-render@document->hw-combobox#rerouteListboxStreamToDialog turbo:before-cache@document->hw-combobox#hideChipsForCache turbo:morph-element->hw-combobox#idempotentConnect keypress.enter->auto-submit#submit"
```

Notice how my extra action has been appended to the end of that string.

However, that won't work because (I believe) before my action of keypress.enter gets the event, this other action runs: `keydown->hw-combobox#navigate`. 
Also, the event of pressing the `Enter` key is bubbling, meaning the it is captured by the innermost element first and then bubbles up all the way to the window.

And that `navigate` action is defined here: https://github.com/josefarias/hotwire_combobox/blob/cf0550098bf8d355872758b7e7ca103ac12c00ae/app/assets/javascripts/hw_combobox/models/combobox/navigation.js#L33

Where the event is "cancelled", which is defined here: https://github.com/josefarias/hotwire_combobox/blob/cf0550098bf8d355872758b7e7ca103ac12c00ae/app/assets/javascripts/hw_combobox/helpers.js#L31

And that stops the propagation of the event, which is why I never get a chance to react to it.

I'm not sure if we can simply forgo calling "cancel" in "navigate", or if it's more complicated than that, and I don't think messing with actions order would be wise. What do you think @josefarias ?
